### PR TITLE
Feat/float support

### PR DESCRIPTION
**Pull request by mmeloni**
this add support for float type.
Please let me know if there are corner cases not yet covered

### CODE REVIEW

File: embedded/sql/catalog.go [hunk 1]
```diff
@@ -404,6 +404,8 @@ func (c *Column) MaxLen() int {
 		return 8
 	case TimestampType:
 		return 8
+	case FloatType:
```
Comment (jeroiraz) [inline embedded/sql/catalog.go:407]: Float64Type ? FloatDoubleType?

File: embedded/sql/engine.go [hunk 2]
```diff
@@ -865,6 +868,22 @@ func EncodeValue(val interface{}, colType SQLValueType, maxLen int) ([]byte, err
...
+		{
+			floatVal, ok := val.(float64)
+			if !ok {
+				return nil, fmt.Errorf(
+					"value is not a float: %w", ErrInvalidValue,
+				)
+			}
+
+			var encv [EncLenLen + 8]byte
+			floatBits := math.Float64bits(floatVal)
```
Comment (jeroiraz) [inline embedded/sql/engine.go:883]: so this encoding preserves lex order with negative values as well, right?

File: embedded/sql/sql_grammar.y [hunk 3]
```diff
@@ -35,7 +35,8 @@ func setResult(l yyLexer, stmts []SQLStmt) {
     values []ValueExp
     value ValueExp
     id string
-    number uint64
+    integer uint64
+	float float64
```
Comment (jeroiraz) [inline embedded/sql/sql_grammar.y:39]: extra space at line 39

File: embedded/sql/stmt.go [hunk 4]
```diff
@@ -1668,6 +1669,75 @@ func (v *Blob) Compare(val TypedValue) (int, error) {
...
+func (v *Float) IsNull() bool {
+	return false
+}
+
+func (v *Float) inferType(cols map[string]ColDescriptor, params map[string]SQLValueType, implicitDB, implicitTable string) (SQLValueType, error) {
+	return FloatType, nil
+}
+
+func (v *Float) requiresType(t SQLValueType, cols map[string]ColDescriptor, params map[string]SQLValueType, implicitDB, implicitTable string) error {
+	if t != TimestampType {
```
Comment (jeroiraz) [inline embedded/sql/stmt.go:1689]: FloatType

Comment (jeroiraz) [reply embedded/sql/stmt.go:1689]: and probably Integer can be converted into Float, but that would be in integer method

File: embedded/sql/stmt.go [hunk 5]
```diff
@@ -1912,6 +1982,8 @@ func (p *Param) substitute(params map[string]interface{}) (ValueExp, error) {
 		{
 			return &Timestamp{val: v}, nil
 		}
+	case float64:
+		return &Float{val: v}, nil
```
Comment (jeroiraz) [inline embedded/sql/stmt.go:1986]: so to maintain same style:

{
     return &Float{val: v}, nil
}

**Review by jeroiraz — COMMENTED** [review]
please add a case for float in method zeroForType in file grouped_row_reader.go

please keep the rest as it's (timestamp is commented), but it's already fixed in another ongoing branch.

**Comment by jeroiraz** [comment]
test cases covering NaN and Infinity may be missing

File: embedded/sql/engine_test.go [hunk 6]
```diff
@@ -341,6 +343,198 @@ func TestTimestampCasts(t *testing.T) {
...
+	err = engine.SetDefaultDatabase("db1")
+	require.NoError(t, err)
+
+	_, _, err = engine.Exec("CREATE TABLE IF NOT EXISTS float_index (id INTEGER AUTO_INCREMENT, ft FLOAT, PRIMARY KEY id)", nil, nil)
+	require.NoError(t, err)
+
+	_, _, err = engine.Exec("CREATE INDEX ON float_index(ft)", nil, nil)
+	require.NoError(t, err)
+
+	floatSerie := []float64{-1.0, 3.345, -0.5, 0.0, -100.8, 0.5, 1.0, math.MaxFloat64, -math.MaxFloat64}
```
Comment (byo) [inline embedded/sql/engine_test.go:514]: Other interesting test cases to check:

* +0.0 vs -0.0 - those can have different representations in floating points although those should end up next to each other in topological sorting order (also to confirm - compiler may convert -0.0 value to 0.0 thus it should be confirmed that that's actually a negative 0 value)
* denormalized values - I think `math.SmallestNonzeroFloat64` should be such value - it should sort correctly compared to 0.0 and other numbers, also a good test would be with greatest positive denormalized value and smallest positive normalized value
* +/- Inf
* +/- NaN

**Review by byo — COMMENTED** [review]
Additional we could add those casts:
* float <-> integer
* float <-> timestamp (this is needed to allow sub-second precision for cast results)

**Comment by byo** [comment]
Missing:

- [ ] immuclient doesn't show float values
- [ ] webconsole support

**Pull request closed without merging**